### ISSUE TITLE
pmdapostgresql: update for postgresql 9.6

### ISSUE THREAD

**Issue by fche** [performancecopilot/pcp#414]
Amazon AWS RDS presents itself as PostgreSQL 9.6.1; Fedora 27 has 9.6.6.  It would be useful if pmdapostgresql could talk to these versions.

**Comment by fche** [comment]
A quick smoke test makes it look like treating 9.6 as 9.5/9.4 is generally OK, as per the attached patch.
[Uploading k.txt…]()

**Comment by goodwinos** [comment]
On both f26 and f27, installed postgresql-9.6.6, applied the patch from the previous comment, built and installed pcp-pmda-postgresql, and then ran PCP qa/778. Then installed postgresql-9.6.7 from the updates-testing repo and ran QA again. All seems fine, so I think it's safe to take the patch - whilst this is only really a smoke test, the qa/778 test does exercise quite a lot of stuff and it looks like 9.6.6 and 9.6.7 are fully backward compatible with 9.4, which is when the PMDA was last tested. I will add to the patch some comments that it's now been tested with 9.6.[67] but not 10.x yet.

Also note: postgresql-10.2 has been released, but is not yet in the Fedora repos. Once that's available, we should audit the PMDA and add any new metrics (10.x is a major release, so there will be changes).

**Comment by natoscott** [comment]
| I think it's safe to take the patch [...]

@goodwinos the patch is not in an even remotely complete state - will discuss offline, cheers.

**Comment by goodwinos** [comment]
OK, so I didn't realize there is a hugely complicated mapping table for the system statistics collector views. Since qa/778 selects all columns from the relevant statistics tables, wouldn't it be failing if there were a discrepancy? In any case, would a reasonable approach here be to run the following script on 9.5 and 9.6, and audit the output to produce the 9.6 mapping table, or at least to verify the statistics tables and views in 9.6 are compatible with 9.5?

```
#! /bin/sh
for table in \
pg_stat_activity \
pg_stat_bgwriter \
pg_stat_database \
pg_stat_all_tables \
pg_stat_sys_tables \
pg_stat_all_indexes \
pg_stat_sys_indexes
do
    echo;echo === TABLE $cmd ===
    sudo -u postgres psql -c "select * from $table"
done

```

**Comment by goodwinos** [comment]
[documentation for statistics in 9.5](https://www.postgresql.org/docs/9.5/static/monitoring-stats.html)
- Table 27-10. pg_stat_all_tables View

[documentation for statistics in 9.6](https://www.postgresql.org/docs/9.6/static/monitoring-stats.html)
- Table 28-12. pg_stat_all_tables View

Those two tables are the same (at least the documentation)

**Comment by natoscott** [comment]
@goodwinos np - and, yep, sounds like a reasonable approach.  Also worth checking the online Postgres docs, which are versioned.  https://www.postgresql.org/docs/9.5/static/monitoring-stats.html

The original patch is basically a throw-away - the patch even says in places it only "may be" correct - so the PMDA needs to be properly audited and updates applied with due diligence and care.  Thanks Mark!

**Comment by goodwinos** [comment]
My audit script (soon to be a QA test) shows there has been a change in the pg_stat_activity table between 9.5.4 and 9.6.7, specifically pg_stat_activity.waiting has been split into two columns as follows 

```
[mgoodwin@f27 postgresql_audit]$ diff postgres_pg_stat_tables.9.5.4 postgres_pg_stat_tables.9.6.7 
1c1
< #version psql (PostgreSQL) 9.5.4
---
> #version psql (PostgreSQL) 9.6.7
16,20c16,21
< 014 pg_stat_activity waiting 
< 015 pg_stat_activity state 
< 016 pg_stat_activity backend_xid 
< 017 pg_stat_activity backend_xmin 
< 018 pg_stat_activity query 
---
> 014 pg_stat_activity wait_event_type 
> 015 pg_stat_activity wait_event 
> 016 pg_stat_activity state 
> 017 pg_stat_activity backend_xid 
> 018 pg_stat_activity backend_xmin 
> 019 pg_stat_activity query 
```

**Comment by goodwinos** [comment]
Completed and merged to master, 0c4e471dc5c5ad and related changes, including unshipped audit tools for developers in src/pmdas/postgresql/audit_tools

**Comment by natoscott** [comment]
@goodwinos great work Mark, thanks for following through & doing this properly.
